ros2/ament.bzl

```diff
@@ -390,3 +391,99 @@ def sh_launcher(name, deps, idl_deps = None, **kwargs):
         ament_setup = ament_setup,
         **kwargs
     )
+
+def _cpp_ament_setup_impl(ctx):
+    src_output = ctx.actions.declare_file(ctx.attr.basename + "/ament_setup.cc")
+    hdr_output = ctx.actions.declare_file(ctx.attr.basename + "/ament_setup.h")
+    ament_prefix_path = ctx.attr.ament_setup[Ros2AmentSetupInfo].ament_prefix_path
+
+    substitutions = {
+        "{{ament_prefix_path}}": ament_prefix_path,
+        "{{header}}": hdr_output.short_path,
+        "{{namespace}}": ctx.attr.basename,
+    }
+
+    expand_template_impl(
+        ctx,
+        template = ctx.file._src_template,
+        output = src_output,
+        substitutions = substitutions,
+        is_executable = False,
+    )
+    expand_template_impl(
+        ctx,
+        template = ctx.file._hdr_template,
+        output = hdr_output,
+        substitutions = substitutions,
+        is_executable = False,
+    )
+
+    files = depset([src_output, hdr_output])
+    return [
+        DefaultInfo(
+            files = files,
+        ),
+    ]
+
+cpp_ament_setup = rule(
+    attrs = {
+        "ament_setup": attr.label(
+            mandatory = True,
+            providers = [Ros2AmentSetupInfo],
+        ),
+        "basename": attr.string(
+            mandatory = True,
+        ),
+        "data": attr.label_list(allow_files = True),
+        "_hdr_template": attr.label(
+            allow_single_file = True,
+            default = "@com_github_mvukov_rules_ros2//ros2:ament_setup.h.tpl",
+        ),
+        "_src_template": attr.label(
+            allow_single_file = True,
+            default = "@com_github_mvukov_rules_ros2//ros2:ament_setup.cc.tpl",
+        ),
+    },
+    implementation = _cpp_ament_setup_impl,
+)
+
+def cpp_ament_setup_library(name, deps, idl_deps = None, **kwargs):
+    """ Generates a C++ library that contains the ament setup code for a ROS2 package.
+
+    This can be useful in contexts where using a wrapper script (as is the case with the regular ros2_xxx rules)
+    is not desirable. The generated function can be called with an optional allow_append argument, which causes the
+    function to append to the AMENT_PREFIX_PATH instead of overwriting it. Beware that when using this mode when
+    different versions of the same plugin exist, the behavior might be unexpected.
```

Comment:
Please add a sentence that the generated cpp function should be called "early enough", where early enough is most likely to be before any ament resource is used.